lib/jsonld-signatures.js

```diff
@@ -266,52 +265,43 @@ api.verify = util.callbackify(async function(input, options) {
   if(options.documentLoader) {
     opts.documentLoader = options.documentLoader;
   }
-  // Why expand just to frame again? The answer is in the proofPurpose
-  // logic. We can't anticipate all the terms we might operate on for all
-  // available ProofPurpose suites, so we expand once and let them
-  // operate on it in expanded form.
-  // (Each ProofPurpose could compact to its own context, though in the case
-  // of ocap-ld, not even which caveats might be used and what the arguments
-  // of them may be is known in advance.)
-  const [expanded] = await jsonld.expand(input, opts);
+
   const framed = await jsonld.compact(
-    expanded,
+    input,
     constants.SECURITY_CONTEXT_URL,
-    Object.assign({}, opts, {skipExpansion: true}));
+    opts);
 
   // ensure there is at least one `proof` or `signature`
   let proofs = jsonld.getValues(framed, 'signature')
     .map(doc => ({property: 'signature', doc}))
     .concat(jsonld.getValues(framed, 'proof')
-      .map(doc => ({property: 'proof', doc})));
+    .map(doc => ({property: 'proof', doc})));
   if(proofs.length === 0) {
     throw new Error('No signature found.');
   }
 
   // Filter proofs to only those with the expected proofpurpose
-  const {purpose} = options;
-  if(purpose) {
-    const {proofPurpose} = purpose;
-    if(!proofPurpose) {
-      throw new Error('`purpose.proofPurpose` is required.');
-    }
-    const ProofPurpose = api.proofPurposes.use(proofPurpose);
-    if(!ProofPurpose) {
-      throw new Error(`Unsupported proof purpose handler "${proofPurpose}".`);
-    }
-    purpose.proofPurposeHandler = new ProofPurpose(injector);
-  }
+  const {purpose, purposeParameters} = options;
+
+  _isSupportedProofPurpose(purpose);
```

Comment:
now that I understand better what's going on here... :)

I think this needs to be moved up and maybe the todo in the helper should be here about throwing in future versions.

We're doing a lot of work before this point just to potentially throw a SyntaxError (missing purpose parameter)

Reply:
> now that I understand better what's going on here... :)
> 
Awesome!
> I think this needs to be moved up and maybe the todo in the helper should be here about throwing in future versions.
> 
> We're doing a lot of work before this point just to potentially throw a SyntaxError (missing purpose parameter)

Okay will do, fix is here: https://github.com/digitalbazaar/jsonld-signatures/pull/39/commits/6ffaf11236bcfa283bd2e1a7b187145340ad9b81